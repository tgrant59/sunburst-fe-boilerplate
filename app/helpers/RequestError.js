class RequestError extends Error {
    constructor(response) {
        super('An error has occurred processing your request')
        this.response = response
    }
}

export default RequestError
