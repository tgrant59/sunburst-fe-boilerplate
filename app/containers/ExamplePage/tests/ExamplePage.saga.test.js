import { put, takeLatest } from 'redux-saga/effects'

import RequestError from 'helpers/RequestError'

import {
    mockCurrentUser,
    mockCurrentUserFirstName,
    mockCurrentUserLastName,
} from './mocks'

import {
    LOAD_USER_REQUESTED,
    loadUserSuccess,
    loadUserFail,
} from '../ExamplePage.actions'
import userSagaWatcher, { getUserData } from '../ExamplePage.saga'

/* eslint-disable redux-saga/yield-effects */
describe('getUserData Saga', () => {
    let getUserDataGenerator

    const mockUserAPIResponse = {
        results: [
            {
                email: mockCurrentUser.email,
                name: {
                    first: mockCurrentUserFirstName,
                    last: mockCurrentUserLastName,
                },
            },
        ],
    }

    beforeEach(() => {
        getUserDataGenerator = getUserData()

        const callDescriptor = getUserDataGenerator.next().value
        expect(callDescriptor).toMatchSnapshot()
    })

    it('should dispatch the reposLoaded action if it requests the data successfully', () => {
        const response = {
            status: 200,
            data: mockUserAPIResponse,
        }
        const putDescriptor = getUserDataGenerator.next(response).value
        const expectedPut = put(loadUserSuccess(mockCurrentUser))
        expect(putDescriptor).toEqual(expectedPut)
    })

    it('should call the repoLoadingError action if the response errors', () => {
        const response = new RequestError({ status: 500 })
        const putDescriptor = getUserDataGenerator.throw(response).value
        const expectedPut = put(loadUserFail(500))
        expect(putDescriptor).toEqual(expectedPut)
    })
})

describe('userSagaWatcher Saga', () => {
    let userSagaWatcherSaga

    beforeEach(() => {
        userSagaWatcherSaga = userSagaWatcher()
    })

    it('should start task to watch for LOAD_USER_REQUESTED action', () => {
        const takeLatestDescriptor = userSagaWatcherSaga.next().value
        const expectedTakeLatest = takeLatest(LOAD_USER_REQUESTED, getUserData)
        expect(takeLatestDescriptor).toEqual(expectedTakeLatest)
    })
})
