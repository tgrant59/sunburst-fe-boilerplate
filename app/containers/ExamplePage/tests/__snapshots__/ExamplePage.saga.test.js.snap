// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserData Saga should call the repoLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://randomuser.me/api/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getUserData Saga should dispatch the reposLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://randomuser.me/api/",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
